### PR TITLE
Basic Product CRUD with User Permissions

### PR DESCRIPTION
**Pull request by aminasaif0**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -46,10 +46,12 @@ gem "bootsnap", require: false
 
 gem "devise"
 
-
+gem "pundit"
 group :development, :test do
   # See https://guides.rubyonrails.org/debugging_rails_applications.html#debugging-with-the-debug-gem
   gem "debug", platforms: %i[ mri windows ]
+  gem "byebug"
```
Comment (salman-haider-confiz) [inline Gemfile:52]: Always lock the Gem version.

File: app/helpers/products_helper.rb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+module ProductsHelper
+end
```
Comment (salman-haider-confiz) [inline app/helpers/products_helper.rb:2]: No Need to create Helper if we are not using it.

File: app/models/concerns/role_check.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+module RoleCheck
+    extend ActiveSupport::Concern
```
Comment (salman-haider-confiz) [inline app/models/concerns/role_check.rb]: indentation issue

File: app/models/role.rb [hunk 4]
```diff
@@ -0,0 +1,4 @@
+class Role < ApplicationRecord
+    has_many :roles_users, dependent: :destroy
```
Comment (salman-haider-confiz) [inline app/models/role.rb]: This table name should be like "user_roles"

File: app/models/user.rb [hunk 5]
```diff
@@ -1,10 +1,13 @@
 class User < ApplicationRecord
   # Include default devise modules. Others available are:
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
+  include RoleCheck
+
   has_many :addresses
   has_many :orders
   has_one :cart
-
+  has_many :roles_users
```
Comment (salman-haider-confiz) [inline app/models/user.rb:9]: same here

File: test/fixtures/roles_users.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (salman-haider-confiz) [inline test/fixtures/roles_users.yml:1]: Remove this file.

File: test/models/role_test.rb [hunk 7]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class RoleTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (salman-haider-confiz) [inline test/models/role_test.rb:7]: Do add test cases as well

File: test/models/roles_user_test.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class RolesUserTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
Comment (salman-haider-confiz) [inline test/models/roles_user_test.rb:7]: Do add testcases as well

File: app/models/user.rb [hunk 9]
```diff
@@ -1,10 +1,13 @@
...
+  has_many :roles, through: :roles_users
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable
 end
```
Comment (salman-haider-confiz) [inline app/models/user.rb]: Do we have any validation check for User table?

**Review by salman-haider-confiz — CHANGES_REQUESTED** [review]

File: app/controllers/products_controller.rb [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    def destroy
+      @product.destroy
+      redirect_to products_url, notice: 'Product was successfully destroyed.'
+    end
+
+    def add_to_cart
+      @product = Product.find(params[:id])
+      if current_user.cart.nil?
+        current_user.create_cart
+      end
```
Comment (saadali7865) [inline app/controllers/products_controller.rb:46]: make it one liner

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+    before_action :authenticate_admin!
+
+    def index
+        @users = User.includes(:roles)
+    end
+    private
+
+  def authenticate_admin!
+    redirect_to root_path, alert: 'Access denied.' unless current_user&.has_role?('admin')
+  end
```
Comment (saadali7865) [inline app/controllers/users_controller.rb:11]: Add space or proper indentation

File: app/helpers/carts_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module CartsHelper
+end
```
Comment (saadali7865) [inline app/helpers/carts_helper.rb:2]: remove this file

File: app/models/concerns/role_check.rb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+module RoleCheck
+  extend ActiveSupport::Concern
+  def has_role?(role)
+    roles.exists?(name: role)
+  end
```
Comment (saadali7865) [inline app/models/concerns/role_check.rb:5]: def is_admin?
  roles.exist?(admin)
end

File: app/policies/application_policy.rb [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+  def edit?
+    update?
+  end
+
+  def destroy?
+    false
+  end
+
+  class Scope
```
Comment (saadali7865) [inline app/policies/application_policy.rb:39]: What is Scope class

File: app/views/layouts/application.html.erb [hunk 15]
```diff
@@ -11,6 +11,22 @@
...
+    <% end %>
+
+    <style type="text/css">
+      .alert-success{
+        color: green;
+      }
+      .alert-danger{
+        color: red;
+      }
+    </style>
```
Comment (saadali7865) [inline app/views/layouts/application.html.erb:28]: Move css to separate file

**Review by saadali7865 — APPROVED** [review]

**Pull request merged**